api/get_users_handler.go

```diff
@@ -0,0 +1,36 @@
+package api
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/alphagov/paas-accounts/database"
+	"github.com/labstack/echo"
+)
+
+func GetUsersHandler(db *database.DB) echo.HandlerFunc {
+	return func(c echo.Context) error {
+		type Users struct {
+			Users  []database.User `json:"users"`
+		}
+
+		params := c.QueryParams()
+		users := Users{}
+
+		if len(params["guids"]) > 0 {
+			users.Users, _ = db.GetUserByUUID(strings.Split(params["guids"][0], ","))
+			return c.JSON(http.StatusOK, users)
+		}
+
+		if params.Get("email") != "" {
+			user, err := db.GetUserByEmail(params.Get("email"))
+			if err != nil {
+				return err
+			}
+		    users.Users = append(users.Users, user)
+			return c.JSON(http.StatusOK, users)
+		}
+
+		return c.JSON(http.StatusOK, users)
```

Comment:
If neither parameter is supplied, we're returning an empty set. As an API consumer, I would expect `/users` to return me a collection of users

Reply:
Or otherwise give me a 4xx error to say I need to supply a query param

Reply:
IMO a `404` is for a _specific_ resource is not found, when the datatype is `Maybe User`, in this case the datatype is `[User]`, if nothing was found it returns `[]`. This reduces the number of states the consumer has to reason about:

```
response = callAPI()

if response.status == 404; return handleErr

users = response.resources

if len(users) == 0; return handleErr

return doStuff(users)
```

Can be reduced to:

```
response = callAPI().resources

if len(users) == 0; return handleErr

return doStuff(users)
```

Reply:
I agree `404` is wrong, but allowing `/users` to return nothing useful without a query string feels like something that would trip people up in the future.

Reply:
Maybe a `400` if the query string is effectively mandatory?